src/bluesky/simulators.py

```diff
@@ -102,5 +117,285 @@ async def check_limits_async(plan):
+        command: str,
+        *,
+        document_name: Optional[str] = None,
+        document: Optional[dict] = None,
+        docs: Optional[Sequence[Sequence[Tuple[str, dict]]]] = None,
+        msg_filter: Optional[Callable[[Msg], bool]] = None,
+    ):
+        """Add a handler to fire one or more callbacks in sequence when a matching command is encountered.
+
+        To fire a single callback just once, specify `document_name` and `document` parameters.
+        To fire a sequence of callbacks, across one or more message invocations, use the `docs` parameter.
+
+        Parameters
+        ----------
+        command
+            Name of the command to match.
+        document_name
+            Name of the document to match when used in single shot mode.
+        document
+            The document to fire in single shot mode.
+        docs: Sequence of Sequence of document_name, document tuples
+            On each receipt of a matching message, the sequence contained in the next element of `docs`
+            will be iterated over and each (document_name, document) pair will be fired to subscribed callbacks
+            in order.
+        """
+        if not docs:
+            if not document_name or not document:
+                raise ValueError("No document sequence specified and document or document_name not specified")
+            docs = [[(document_name, document)]]
+        it = iter(docs)
+
+        def handle_command(_):
+            for name, doc in next(it):
+                self.fire_callback(name, doc)
+
+        self.add_handler(command, handle_command, msg_filter)
+
+    def fire_callback(self, document_name, document) -> None:
+        """Fire all the callbacks registered for this document type in order to simulate something happening.
+
+        Parameters
+        ----------
+        document_name
+            document name as defined in the Bluesky Message Protocol 'subscribe' call,
+            all subscribers filtering on this document name will be called
+        document
+            the document to send
+        """
+        for callback_func, callback_docname in self.callbacks.values():
+            if callback_docname == "all" or callback_docname == document_name:
+                callback_func(document_name, document)
+
+    def simulate_plan(self, gen: Generator[Msg, Any, Any]) -> List[Msg]:
+        """Simulate the RunEngine executing the plan.
+
+        After executing the plan return_value is populated with the return value of the plan.
+
+        Parameters
+        ----------
+        gen
+            the generator function that executes the plan
+
+        Returns
+        -------
+        list[Msg]
+            list of the messages generated by the plan
+        """
+        messages = []
+        send_value = None
+        try:
+            while msg := gen.send(send_value):
+                send_value = None
+                messages.append(msg)
+                LOGGER.debug(f"<{msg}")
```

Comment:
Best practice is to let the logging framework do the templating to avoid paying the cost of generating the string if the logging level is such that this log never gets emitted. 
```suggestion
                LOGGER.debug("<{%s}", msg)
```

(and similar below)